doc/HLD_of_FOL.rst

```diff
@@ -0,0 +1,290 @@
+============================================
+High level design of a file operations log
+============================================
+
+This document presents a high level design (HLD) of a file operations log (a fol) of Motr M0 core. The main purposes of this document are: (i) to be inspected by M0 architects and peer designers to ascertain that high level design is aligned with M0 architecture and other designs, and contains no defects, (ii) to be a source of material for Active Reviews of Intermediate Design (ARID) and detailed level design (DLD) of the same component, (iii) to serve as a design reference document.
+
+The intended audience of this document consists of M0 customers, architects, designers and developers.
+
+*************
+Introduction
+*************
+
+A fol is a central M0 data-structure, maintained by every node where M0 core is deployed and serving multiple goals:
+
+- it is used by a node data-base component to implement local transactions through WAL logging;
+
+- it is used by DTM to implement distributed transactions. DTM uses fol for multiple purposes internally:
+
+  - on a node where a transaction originates: for replay;
+
+  - on a node where a transaction update is executed: for undo, redo and for replay (sending redo requests to recovering nodes);
+
+  - to determine when a transaction becomes stable;
+
+- it is used by a cache pressure handler to determine what cached updates have to be re-integrated into upward caches;
+
+- it is used by FDML to feed file system updates to fol consumers asynchronously;
+
+- more generally, a fol is used by various components (snapshots, addb, etc.) to consistently reconstruct the system state as at a certain moment in the (logical) past.
+
+Roughly speaking, a fol is a partially ordered collection of fol records, each corresponding to (part of) a consistent modification of file system state. A fol record contains information determining durability of the modification (how many volatile and persistent copies it has and where, etc.) and dependencies between modifications, among other things. When a client node has to modify a file system state to serve a system call from a user, it places a record it its (possibly volatile) fol. The record keeps track of operation state: has it been re-integrated to servers, has it been committed on the servers, etc. A server, on receiving a request to execute an update on a client behalf, inserts a record, describing the request into its fol. Eventually, fol is purged to reclaim storage, culling some of the records.
```

Comment:
s
/it places a record it its (possibly volatile) fol
/it places a record in its (possibly volatile) fol

Reply:
Incorporated.